common/src/main/java/io/netty/util/concurrent/Future.java

```diff
@@ -306,6 +307,52 @@
     @Override
     boolean cancel(boolean mayInterruptIfRunning);
 
+    /**
+     * Create a new future that will complete with the result of this future mapped through the given mapper function.
+     * <p>
+     * If this future fails, then the returned future will fail as well, with the same exception.
+     * Cancellation of either future will cancel the other.
+     * If the mapper function throws, the returned future will fail, but this future will be unaffected.
+     *
+     * @param mapper The function that will convert the result of this future into the result of the returned future.
+     * @param <R> The result type of the mapper function, and of the returned future.
+     * @return A new future instance that will complete with the mapped result of this future.
+     */
+    default <R> Future<R> map(Function<V, R> mapper) {
```

Comment:
I think we should not mix methods that mutate the `Future` instance like `addListener()` and methods that decorates and returns a new `Future` instance. It may be obvious with `map()`/`flatMap()` as they change the generic type but with operators like `filter()` it muddies the water.

Instead, a static class like `Futures` may be fine to provide all operators.

Reply:
How would a `filter()` method work? A future is not really a collection or a stream.